### ISSUE TITLE
[java] rulesets/java/typeresolution.xml lists non-existent rules

### ISSUE THREAD

**Issue by asarkar** [pmd/pmd#866]
**Affects PMD Version:** 
6.0.0

**Rule:**
rulesets/java/typeresolution.xml

**Description:**
The following rules don't exist.

```
# this is actually in category/java/bestpractices.xml
<rule ref="java-coupling/LooseCoupling" deprecated="true" />
# this is actually in category/java/errorprone.xml
<rule ref="java-clone/CloneMethodMustImplementCloneable" deprecated="true" />
# this is actually in category/java/bestpractices.xml
<rule ref="java-imports/UnusedImports" deprecated="true" />
# this is actually in category/java/design.xml
<rule ref="java-strictexception/SignatureDeclareThrowsException" deprecated="true" />
```

**Running PMD through:** *[Gradle + Other]*

**Comment by jsotuyod** [comment]
@asarkar those are references to the old rulesets.

`java-coupling/LooseCoupling` is equivalent to `rulesets/coupling.xml/LooseCoupling`. That rule, in turn, is defined as:

```xml
<rule ref="category/java/bestpractices.xml/LooseCoupling" deprecated="true" />
```

This is down to:
1. `typeresolution`-specific implementations no longer exist. They "graduated" to full implementations, replacing the old "non-typeresolution" one, so we reference those.
2. the other rulesets got deprecated, with all rules being moved to categories.

Loading those rules will work (with proper log messages).

I guess this is being reported as you are having some issues with your migration tool? That's probably just lack of support for `java-ruleset` format of references. You can do that with any ruleset.

Are you having any trouble with this behaviour?

**Comment by asarkar** [comment]
> the other rulesets got deprecated, with all rules being moved to categories.

Correct, and I've noted in the defect where they ended up. 

> I guess this is being reported as you are having some issues with your migration tool

Correct again, that's how I found out that the locations are invalid; the other rulesets correctly list the new location, whereas this one doesn't. Seems like a small oversight to me.

**Comment by jsotuyod** [comment]
> Correct again, that's how I found out that the locations are invalid; the other rulesets correctly list the new location, whereas this one doesn't. Seems like a small oversight to me.

It's not really. You will have the same problem if a user gives you an input ruleset such as:

```xml
<ruleset xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" name="Android Application Rules"
 xmlns="http://pmd.sf.net/ruleset/1.0.0"
 xsi:noNamespaceSchemaLocation="http://pmd.sf.net/ruleset_xml_schema.xsd"
 xsi:schemaLocation="http://pmd.sf.net/ruleset/1.0.0 http://pmd.sf.net/ruleset_xml_schema.xsd">
   <description>PMD Ruleset</description>
   <rule ref="java-coupling"/>
   <rule ref="java-empty"/>
</ruleset>
```

This is completely valid and will work on PMD. Your tool, therefore, should be able to process it for compatibility.

You can see how these references are parsed / resolved here:

https://github.com/pmd/pmd/blob/c6bc5ff30672e43848923761a8e2627d99bdf656/pmd-core/src/main/java/net/sourceforge/pmd/RuleSetReferenceId.java#L247-L266

**Comment by jsotuyod** [comment]
@asarkar I can remove the indirection from `typeresolution.xml` if it helps you (removes recursive resolution), but you would still need to understand that reference format.

Also, you should check to properly support things such as:

```xml
<ruleset xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" name="Android Application Rules"
 xmlns="http://pmd.sf.net/ruleset/1.0.0"
 xsi:noNamespaceSchemaLocation="http://pmd.sf.net/ruleset_xml_schema.xsd"
 xsi:schemaLocation="http://pmd.sf.net/ruleset/1.0.0 http://pmd.sf.net/ruleset_xml_schema.xsd">
   <description>PMD Ruleset</description>
   <rule ref="java-coupling">
     <exclude name="LooseCoupling"/> <!-- exclude LooseCoupling -->
   </rule>
   <rule ref="java-typeresolution"/> <!-- includes LooseCoupling -->
</ruleset>
```

Not sure if this scenario is properly handled right now

**Comment by asarkar** [comment]
Interesting. Based on my browsing in the repo, I assumed all references as path relative to https://github.com/pmd/pmd/tree/master/pmd-java/src/main/resources. The code snippet you provided is more involved, and I can easily accommodate it. I don't think any action is necessary on your side.

**Comment by asarkar** [comment]
Note that recursion wasn't an issue to begin with, I do resolve rulesets referencing to other rulesets. 
https://github.com/asarkar/pmd-migration-tool/blob/master/src/main/kotlin/org/abhijitsarkar/kotlin/pmd/Migrator.kt#L146
It's the path that was throwing me off.
